### PR TITLE
add cache invalidate field

### PR DESCRIPTION
**Pull request by ETLaurent**
Part of [PRO-2576](https://linear.app/apostrophecms/issue/PRO-2576/caching-emit-etags-for-the-context-document-and-invalidate-them-based)

## Summary

- Add a `cacheInvalidatedAt` field via a migration for docs that don't have it yet, and set it to equal their `updatedAt` field value
- Set `cacheInvalidatedAt` value to the current date when saving a doc

## What are the specific steps to test this change?

- Run the website

👉 `add-cache-invalidated-at-field-for-*` migration should be done only once:

```
Running database migration: add-cache-invalidated-at-field
```

👉 `cacheInvalidatedAt` field should have been added to all documents from the `aposDocs` collection, and set to equal the docs' `updatedAt` field.

- Add a piece, or a page

👉 The document created (piece or page for instance) should have a `cacheInvalidatedAt` that equals the `updatedAt` field.

## What kind of change does this PR introduce?
*(Check at least one)*

- [ ] Bug fix
- [x] New feature
- [ ] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other

## Make sure the PR fulfills these requirements:

- [x] It includes a) the existing issue ID being resolved, b) a convincing reason for adding this feature, or c) a clear description of the bug it resolves
- [ ] The changelog is updated
- [ ] Related documentation has been updated
- [x] Related tests have been updated

If adding a new feature without an already open issue, it's best to open a **feature request issue** first and wait for approval before working on it.

### CODE REVIEW

File: test/page-type.js [hunk 1]
```diff
@@ -20,13 +21,27 @@ describe('page-type', function() {
...
+          }
         }
       }
     });
 
     assert(apos && apos.__meta.name === 'apostrophe');
   });
 
+  it('should add the migration that sets the cache field', () => {
+    assert(migrations.includes('add-cache-invalidated-at-field-for-nifty-page'));
```
Comment (ETLaurent) [inline test/page-type.js:42]: very sketchy...

File: modules/@apostrophecms/doc-type/index.js [hunk 2]
```diff
@@ -502,6 +502,23 @@ module.exports = {
         }
         return self.apos.migration.addSortify(self.__meta.name, { type: self.name }, field);
       },
+      // Add the "cacheInvalidatedAt" field to the documents that do not have it yet,
+      // and set it to equal doc.updatedAt.
+      // The migration will apply only for the documents of the same type as
+      // the modules' that use this method, via `self.apos.migration.add`.
+      addCacheMigration() {
+        self.apos.migration.eachDoc({ type: self.__meta.name }, 5, async doc => {
```
Comment (ETLaurent) [inline modules/@apostrophecms/doc-type/index.js:510]: should we exclude the `@apostrophecms/*` pieces?

Comment (ETLaurent) [reply modules/@apostrophecms/doc-type/index.js:510]: should we exclude drafts?

File: modules/@apostrophecms/doc-type/index.js [hunk 3]
```diff
@@ -502,6 +502,23 @@ module.exports = {
...
+          await self.apos.doc.db.updateOne(
+            {
+              _id: doc._id,
+              cacheInvalidatedAt: { $exists: 0 }
```
Comment (ETLaurent) [inline modules/@apostrophecms/doc-type/index.js:514]: should we exclude drafts?

File: modules/@apostrophecms/doc-type/index.js (hunk 2 above)
Comment (boutell) [reply modules/@apostrophecms/doc-type/index.js:510]: No, there's no harm in being consistent and someone else might decide to make a piece-page for them.

File: modules/@apostrophecms/doc-type/index.js [hunk 4]
```diff
@@ -883,6 +900,7 @@ module.exports = {
         // Setting it this way rather than setting it to published.updatedAt
         // guarantees no small discrepancy breaking equality comparisons
         draft.updatedAt = draft.lastPublishedAt;
+        draft.cacheInvalidatedAt = draft.lastPublishedAt;
```
Comment (boutell) [inline modules/@apostrophecms/doc-type/index.js:903]: 👍

**Review by boutell — CHANGES_REQUESTED** [review]

File: modules/@apostrophecms/doc-type/index.js [hunk 5]
```diff
@@ -502,6 +503,23 @@ module.exports = {
         }
         return self.apos.migration.addSortify(self.__meta.name, { type: self.name }, field);
       },
+      // Add the "cacheInvalidatedAt" field to the documents that do not have it yet,
+      // and set it to equal doc.updatedAt.
+      addCacheFieldMigration() {
+        self.apos.migration.add(`add-cache-invalidated-at-field-for-${self.__meta.name}`, () => {
```
Comment (boutell) [inline modules/@apostrophecms/doc-type/index.js:509]: That's a lot of migrations and the behavior doesn't differ from one doc type to another. So write just one migration in `@apostrophecms/doc` which operates on all docs.

File: modules/@apostrophecms/doc-type/index.js [hunk 6]
```diff
@@ -502,6 +503,23 @@ module.exports = {
...
+          self.apos.migration.eachDoc({ type: self.__meta.name }, 5, async doc => {
```
Comment (boutell) [inline modules/@apostrophecms/doc-type/index.js:510]: This is the right place to do `cacheInvalidatedAt: { $exists: 0 }`. Doing that in `updateOne` means we pay for an update call for every doc, whether it already has this property or not.

**Review by boutell — CHANGES_REQUESTED** [review]

File: test/docs.js [hunk 7]
```diff
@@ -30,6 +35,37 @@ describe('Docs', function() {
...
+        '@apostrophecms/migration': {
+          extendMethods(self) {
+            return {
+              add(_super, name, migrationFn, options) {
+                migrations.addList.push(name);
+                migrationFn();
+                return _super(name, migrationFn, options);
+              },
+              eachDoc(_super, criteria, limit, iterator) {
+                if (criteria.cacheInvalidatedAt) {
```
Comment (boutell) [inline test/docs.js:48]: This is going to alter the behavior of every eachDoc call in every migration, which could be quite a lot of unexpected behavior.

Also your `add` call adds a migration for every other migration.

You should call `self.apos.migration.add` in one of the modules of your test, i.e. call that from `init` of that module or some method invoked by `init`. 

Here is an example:

https://github.com/apostrophecms/apostrophe/blob/main/modules/%40apostrophecms/area/index.js#L581-L606

Comment (boutell) [reply test/docs.js:48]: Your migration will call the regular `eachDoc` method. There is no need to alter the behavior of all `eachDoc` methods for every migration, unless I am missing something.

File: test/docs.js [hunk 8]
```diff
@@ -30,6 +35,37 @@ describe('Docs', function() {
...
+                  const _updateOne = self.apos.doc.db.updateOne.bind(self.apos.doc.db);
+                  self.apos.doc.db.updateOne = (filter, update, options) => {
```
Comment (boutell) [inline test/docs.js:50]: This is a really big override of every mongo update call to add migrations. There is definitely no need for that. Please reach out if you have questions after reviewing some of the regular migrations in apostrophe core.

**Review by boutell — CHANGES_REQUESTED** [review]

File: modules/@apostrophecms/doc/index.js [hunk 9]
```diff
@@ -1093,6 +1096,17 @@ module.exports = {
           }
         });
       },
+      // Add the "cacheInvalidatedAt" field to the documents that do not have it yet,
+      // and set it to equal doc.updatedAt.
+      addCacheFieldMigration() {
+        self.apos.migration.add('add-cache-invalidated-at-field', () => {
```
Comment (boutell) [inline modules/@apostrophecms/doc/index.js:1102]: 👍

File: test/docs.js (hunk 7 above)
Comment (boutell) [reply test/docs.js:48]: I see that you wrote a new migration which is correct, but this code in the tests to override all migrations in a strange way is still here. I think you meant to remove it?

**Review by boutell — CHANGES_REQUESTED** [review]

**Review by boutell — APPROVED** [review]

**Pull request merged**